File: ios/SecretaryDeviceUnlock/Sources/SecretaryDeviceUnlock/DeviceUnlockCoordinator.swift
import Foundation

/// Pure orchestration over three injected ports. No I/O of its own.
///
/// Explicitly `Sendable` (its three ports are `Sendable`): a non-frozen public
/// struct does not export synthesized `Sendable` across the module boundary, and
/// a `@MainActor` view model sends it off-actor to call the `async` `unlock`
/// (#231).
public struct DeviceUnlockCoordinator: Sendable {
    let slotPort: VaultDeviceSlotPort
    let enclave: DeviceSecretEnclave
    let metadata: DeviceEnrollmentMetadataStore

    public init(slotPort: VaultDeviceSlotPort,
                enclave: DeviceSecretEnclave,
                metadata: DeviceEnrollmentMetadataStore) {
        self.slotPort = slotPort
        self.enclave = enclave
        self.metadata = metadata
    }

    /// Enroll this device: password-open + mint a slot, SE-wrap the secret,
    /// persist metadata. Transactional: any mid-flow failure rolls back so no
    /// orphan wrap file or enclave key survives.
    public func enroll(vaultPath: Data, vaultId: String, password: [UInt8]) throws {
        var slot = try mapSlotErrors { try slotPort.addDeviceSlot(vaultPath: vaultPath, password: password) }
        defer { zeroize(&slot.deviceSecret) }

        do {
            try enclave.store(secret: slot.deviceSecret)
        } catch {
            try? slotPort.removeDeviceSlot(vaultPath: vaultPath, deviceUuid: slot.deviceUuid)
            throw error
        }

        do {
            try metadata.save(DeviceEnrollment(vaultId: vaultId, deviceUuid: slot.deviceUuid))
        } catch {
            // Roll back best-effort and rethrow the ORIGINAL save error. The
            // rollback failures are swallowed (`try?`) deliberately: enrollment
            // already did not complete, so a residual SE key / wrap file is
            // recoverable — the next enroll's `store` replaces the key and mints
            // a fresh slot — and masking the real cause with a rollback error
            // would be worse than leaving a recoverable remnant.
            try? enclave.clear()
            try? slotPort.removeDeviceSlot(vaultPath: vaultPath, deviceUuid: slot.deviceUuid)
            throw error
        }
    }

    /// The most recent release diagnostic from the enclave (raw domain+code on a
    /// biometric/decrypt failure), for a UI to surface. Read immediately after a
    /// failed `unlock`; nil after a successful release.
    public var lastReleaseDiagnostic: String? { enclave.lastReleaseDiagnostic }

    /// True iff both the enclave holds a secret AND enrollment metadata exists.
    /// Non-throwing by contract, so a metadata read error reads as "not enrolled"
    /// (`try?` flattens both a thrown error and a nil load to `nil`).
    public var isEnrolled: Bool {
        enclave.isEnrolled && (try? metadata.load()) != nil
    }

    /// Disenroll this device: remove the vault slot (tolerating an already-gone
    /// slot), then clear the enclave key and metadata so no orphan survives.
    public func disenroll(vaultPath: Data) throws {
        if let enrollment = try metadata.load() {
            do {
                try slotPort.removeDeviceSlot(vaultPath: vaultPath, deviceUuid: enrollment.deviceUuid)
            } catch VaultSlotError.deviceSlotNotFound {
                // already gone — fine
            }
        }
        try enclave.clear()
        try metadata.clear()
    }

    /// Biometric-release the device secret for the enrolled vault WITHOUT opening
    /// the vault. Metadata guard runs BEFORE the enclave prompt (no prompt when
    /// not enrolled). The caller zeroizes `credential.secret` after the open.
    public func releaseCredential(reason: String) async throws -> DeviceSecretCredential {
        guard let enrollment = try metadata.load() else { throw DeviceUnlockError.notEnrolled }
        let secret = try await enclave.release(reason: reason) // throws DeviceUnlockError
        return DeviceSecretCredential(deviceUuid: enrollment.deviceUuid,
                                      secret: secret,
                                      enrolledVaultId: enrollment.vaultId)
    }

    /// Unlock: biometric-release the secret (via `releaseCredential`), then open
    /// the vault with it. Retains the `vaultId` guard so a stale enrollment for a
    /// different vault fails BEFORE the biometric prompt.
    public func unlock(vaultPath: Data, vaultId: String, reason: String) async throws -> OpenedVault {
        guard let enrollment = try metadata.load() else { throw DeviceUnlockError.notEnrolled }
        guard enrollment.vaultId == vaultId else { throw DeviceUnlockError.vaultSlotMismatch }

        var cred = try await releaseCredential(reason: reason)
        defer { zeroize(&cred.secret) }

        return try mapSlotErrors {
            try slotPort.openWithDeviceSecret(vaultPath: vaultPath,
                                              deviceUuid: cred.deviceUuid,
                                              deviceSecret: cred.secret)
        }
    }
}

/// Run a port call, translating `VaultSlotError` into the coordinator's typed
/// `DeviceUnlockError` semantics. Module-internal: `unlock` reuses it, and
/// `enroll`'s add path re-throws the mapped error through the caller's `try`.
internal func mapSlotErrors<R>(_ body: () throws -> R) throws -> R {
    do {
        return try body()
    } catch let e as VaultSlotError {
        switch e {
        case .deviceSlotNotFound:          throw DeviceUnlockError.vaultSlotMismatch
        case .wrongDeviceSecretOrCorrupt:  throw DeviceUnlockError.wrongDeviceSecretOrCorrupt
        default:                           throw DeviceUnlockError.vault(e)
        }
    }
}

File: android/vault-access/src/main/kotlin/org/secretary/browse/DeviceUnlockCoordinator.kt
package org.secretary.browse

/**
 * Pure orchestration of device enrollment / unlock / disenroll over three injected ports. No I/O of
 * its own. Mirror of iOS `DeviceUnlockCoordinator`, with ONE deliberate divergence: [unlock] returns
 * an [UnlockCredential.DeviceSecret] (the caller opens via `openWithCredential`) rather than opening
 * directly — so open-time errors surface as [VaultBrowseError] from the shared pipeline, consistent
 * with the password/recovery paths.
 */
class DeviceUnlockCoordinator(
    private val slotPort: VaultDeviceSlotPort,
    private val enclave: DeviceSecretEnclave,
    private val metadata: DeviceEnrollmentMetadataStore,
) {
    /** True iff BOTH the enclave holds a secret AND enrollment metadata is present. */
    val isEnrolled: Boolean
        get() = enclave.isEnrolled && runCatching { metadata.load() }.getOrNull() != null

    /**
     * Mint a device slot, store its secret in the enclave, and record the enrollment — transactionally.
     * On enclave-store failure the slot is removed; on metadata-save failure both the enclave and the
     * slot are rolled back and the ORIGINAL save error is rethrown. The slot's secret copy is zeroized
     * on every exit. [password] is owned by the caller (forwarded to `addDeviceSlot`, not zeroized here).
     */
    suspend fun enroll(folder: String, vaultId: String, password: ByteArray) {
        val slot = slotPort.addDeviceSlot(folder, password)
        try {
            try {
                enclave.store(slot.secret)
            } catch (e: Throwable) {
                runCatching { slotPort.removeDeviceSlot(folder, slot.deviceUuid) }
                throw e
            }
            try {
                metadata.save(DeviceEnrollment(vaultId, slot.deviceUuid))
            } catch (e: Throwable) {
                runCatching { enclave.clear() }
                // A process crash between enclave.clear() and removeDeviceSlot() below can leave an
                // orphan devices/<uuid>.wrap with no enclave secret and no metadata — benign (effectively
                // dead; idempotent disenroll cleans it).
                runCatching { slotPort.removeDeviceSlot(folder, slot.deviceUuid) }
                throw e
            }
        } finally {
            slot.secret.fill(0)
        }
    }

    /**
     * Release the device secret (slice 2: behind a biometric prompt) and wrap it into an
     * [UnlockCredential.DeviceSecret]. Guards run BEFORE [DeviceSecretEnclave.release] so a stale /
     * wrong-vault enrollment never triggers a biometric prompt: [DeviceUnlockError.NotEnrolled] if no
     * metadata, [DeviceUnlockError.VaultSlotMismatch] if the enrolled vaultId differs. The returned
     * credential owns the secret; the CALLER opens via `openWithCredential` and zeroizes it.
     */
    suspend fun unlock(vaultId: String, reason: String): UnlockCredential.DeviceSecret {
        val enrollment = metadata.load() ?: throw DeviceUnlockError.NotEnrolled
        if (enrollment.vaultId != vaultId) throw DeviceUnlockError.VaultSlotMismatch
        val secret = enclave.release(reason)
        return UnlockCredential.DeviceSecret(enrollment.deviceUuid, secret)
    }

    /**
     * Revoke this device's enrollment, idempotently. Removes the slot (a
     * [VaultBrowseError.DeviceSlotNotFound] is swallowed — already-gone is success; any other
     * [VaultBrowseError] propagates), then best-effort clears the enclave + metadata. Safe when not
     * enrolled (nothing to remove). No orphan survives.
     */
    suspend fun disenroll(folder: String) {
        val enrollment = metadata.load()
        if (enrollment != null) {
            try {
                slotPort.removeDeviceSlot(folder, enrollment.deviceUuid)
            } catch (e: VaultBrowseError.DeviceSlotNotFound) {
                // already gone — fine
            }
        }
        runCatching { enclave.clear() }
        runCatching { metadata.clear() }
    }
}
